provider/image_test.go

```diff
@@ -0,0 +1,300 @@
+package provider
+
+import (
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestSetRegistry(t *testing.T) {
+
+	t.Run("Valid Registry", func(t *testing.T) {
+		expected := Registry{
+			Server:   "https://index.docker.io/v1/",
+			Username: "pulumipus",
+			Password: "supersecret",
+		}
+		input := resource.NewObjectProperty(resource.PropertyMap{
+			"server":   resource.NewStringProperty("https://index.docker.io/v1/"),
+			"username": resource.NewStringProperty("pulumipus"),
+			"password": resource.NewStringProperty("supersecret"),
+		})
+
+		actual := setRegistry(input)
+		assert.Equal(t, expected, actual)
+	})
+	t.Run("Incomplete Registry sets all available fields", func(t *testing.T) {
+		expected := Registry{
+			Server:   "https://index.docker.io/v1/",
+			Username: "pulumipus",
+		}
+		input := resource.NewObjectProperty(resource.PropertyMap{
+			"server":   resource.NewStringProperty("https://index.docker.io/v1/"),
+			"username": resource.NewStringProperty("pulumipus"),
+		})
+
+		actual := setRegistry(input)
+		assert.Equal(t, expected, actual)
+	})
+
+	t.Run("Registry can be nil", func(t *testing.T) {
+		expected := Registry{}
+		input := resource.PropertyValue{}
+		actual := setRegistry(input)
+		assert.Equal(t, expected, actual)
+	})
+}
+
+func TestMarshalBuild(t *testing.T) {
+
+	t.Run("Default Build on empty input", func(t *testing.T) {
+		expected := Build{
+			Context:    ".",
+			Dockerfile: "Dockerfile",
+		}
+		input := resource.PropertyValue{
+			resource.NewPropertyMapFromMap(map[string]interface{}{}),
+		}
+		actual := marshalBuild(input)
```

Comment:
Clearly from this test it looks like marshalBuild applies default values. This is a bit unexpected given the name "marshalBuild" the "marshal" part makes me expect something that just does 'boring' conversions between formats. Perhaps "marshalBuildAndApplyDefaults"?